pypesto/hierarchical/optimal_scaling_approach/optimal_scaling_calculator.py

```diff
@@ -0,0 +1,170 @@
+        amici_model:
+            The AMICI model.
+        amici_solver:
+            The AMICI solver.
+        edatas:
+            The experimental data.
+        n_threads:
+            Number of threads for AMICI call.
+        x_ids:
+            Ids of optimization parameters.
+        parameter_mapping:
+            Mapping of optimization to simulation parameters.
+        fim_for_hess:
+            Whether to use the FIM (if available) instead of the Hessian (if
+            requested).
+        """
+        if mode == MODE_RES:
+            raise ValueError(
+                "Optimal scaling method cannot be called with residual mode."
+            )
+
+        # get dimension of outer problem
+        dim = len(x_ids)
+
+        # initialize return values
+        nllh, snllh, s2nllh, chi2, res, sres = init_return_values(
+            sensi_orders, mode, dim
+        )
+        # set order in solver
+        sensi_order = 0
+        if sensi_orders:
+            sensi_order = max(sensi_orders)
+
+        amici_solver.setSensitivityOrder(sensi_order)
+
+        x_dct = copy.deepcopy(x_dct)
+
+        # fill in parameters
+        amici.parameter_mapping.fill_in_parameters(
+            edatas=edatas,
+            problem_parameters=x_dct,
+            scaled_parameters=True,
+            parameter_mapping=parameter_mapping,
+            amici_model=amici_model,
+        )
+        # run amici simulation
+        inner_rdatas = amici.runAmiciSimulations(
+            amici_model,
+            amici_solver,
+            edatas,
+            num_threads=min(n_threads, len(edatas)),
+        )
+        inner_result = {
+            FVAL: nllh,
+            GRAD: snllh,
+            HESS: s2nllh,
+            RES: res,
+            SRES: sres,
+            RDATAS: inner_rdatas,
+            X_INNER_OPT: self.inner_problem.get_inner_parameter_dictionary(),
+        }
+
+        # if any amici simulation failed, it's unlikely we can compute
+        # meaningful inner parameters, so we better just fail early.
+        if any(rdata.status != amici.AMICI_SUCCESS for rdata in inner_rdatas):
+            # if the gradient was requested, we need to provide some value
+            # for it
+            inner_result[FVAL] = np.inf
+            if 1 in sensi_orders:
+                inner_result[GRAD] = np.full(
+                    shape=len(x_ids), fill_value=np.nan
+                )
+            return filter_return_dict(inner_result)
+
+        sim = [rdata['y'] for rdata in inner_rdatas]
+
+        # compute optimal inner parameters
+        x_inner_opt = self.inner_solver.solve(self.inner_problem, sim)
+        inner_result[FVAL] = self.inner_solver.calculate_obj_function(
+            x_inner_opt
+        )
+        inner_result[
+            X_INNER_OPT
+        ] = self.inner_problem.get_inner_parameter_dictionary()
+
+        # calculate analytical gradients if requested
+        if sensi_order > 0:
```

Comment:
check if sensi_order == 2 (Hessian) and fail? As FIM cannot be calculated. (Fail at beginning of function)